File: backend/agentforge/forge_client.py
"""forge_markets on-chain client — build the four instructions, first-call-correct.

Turns an intent (open a market, stake a side, settle with a proof, claim) into the EXACT
``forge_markets`` instruction the deployed devnet program expects: right 8-byte Anchor
discriminator, right account metas (order + signer/writable flags), right Borsh args, right
PDA derivations. The program (``program/programs/forge-markets``) is frozen; this module
mirrors its wire format as DATA, the same duplication pattern the Rust test crate uses.

Nothing here signs or sends — it only *builds*. The policy-gated wallet
(:mod:`agentforge.wallets`) signs an :class:`UnsignedTx`; :mod:`agentforge.solana_rpc` moves
bytes. Keeping build/sign/send separate is what lets the whole thing be falsified offline.

The settle path carries the probe's hard-won fixes:
  * ``ts`` is ``summary.updateStats.minTimestamp`` (the oracle's daily-root seed), NOT the
    proof's top-level ``ts`` (else error 6010 TimestampMismatch inside the oracle);
  * the ``daily_scores_roots`` PDA is derived from that same ``minTimestamp``;
  * a ComputeBudget ``SetComputeUnitLimit`` prelude (~350k) — the ``validate_stat`` CPI
    burns ~205k;
  * settle carries NO predicate (the program injects the market's stored predicate into the
    CPI — the caller cannot override the condition being proven).
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from enum import IntEnum
from pathlib import Path
from typing import Any

from solders.instruction import AccountMeta, Instruction
from solders.pubkey import Pubkey

from .decision import Side

# ── program identities (frozen; verified deployed + executable on devnet) ────────
FORGE_PROGRAM_ID = Pubkey.from_string("7Pvo6SEh1zBa1Euvj5QQ4td9GpfsQosTpxhqwWtWUFt6")
TXORACLE_PROGRAM_ID = Pubkey.from_string("6pW64gN1s2uqjHkn1unFeEjAwJkPGHoppGvS715wyP2J")
SYSTEM_PROGRAM_ID = Pubkey.from_string("11111111111111111111111111111111")
COMPUTE_BUDGET_ID = Pubkey.from_string("ComputeBudget111111111111111111111111111111")

# ── PDA seeds (mirror forge-markets/src/interface.rs) ────────────────────────────
MARKET_SEED = b"market"
VAULT_SEED = b"vault"
POSITION_SEED = b"position"
ROOTS_SEED = b"daily_scores_roots"  # txoracle-owned daily Merkle roots

# ── Anchor discriminators = sha256("global:<ix>")[..8] (verified against the IDL) ─
DISCRIMINATORS: dict[str, bytes] = {
    "create_market": bytes([103, 226, 97, 235, 200, 188, 251, 254]),
    "stake": bytes([206, 176, 202, 18, 200, 209, 179, 108]),
    "settle": bytes([175, 42, 185, 87, 144, 131, 102, 212]),
    "claim": bytes([62, 198, 214, 193, 213, 159, 108, 210]),
}

# validate_stat burns ~205k CU inside the CPI; give settle head-room.
SETTLE_COMPUTE_UNITS = 350_000

# Programs an on-chain wallet tolerates as a non-spending prelude to a whitelisted call.
PRELUDE_PROGRAM_IDS = frozenset({COMPUTE_BUDGET_ID})

_EPOCH_DAY_MS = 86_400_000


class ForgeError(Exception):
    """Building a forge_markets instruction failed (bad proof shape, out-of-range value)."""


class Comparison(IntEnum):
    """txoracle ``Comparison`` — Borsh variant index (single source of truth)."""

    GREATER_THAN = 0
    LESS_THAN = 1
    EQUAL_TO = 2


# Borsh variant index for the program's ``Side`` enum (Yes = 0, No = 1).
_SIDE_INDEX: dict[Side, int] = {"Yes": 0, "No": 1}


def side_index(side: Side) -> int:
    return _SIDE_INDEX[side]


def to_lamports(sol: float) -> int:
    """SOL (the WalletSeam's float unit) → lamports (the wire unit)."""
    return int(round(sol * 1_000_000_000))


@dataclass(frozen=True)
class TraderPredicate:
    """The YES condition a market stores and txoracle later evaluates. Borsh: i32 + u8."""

    threshold: int
    comparison: Comparison


@dataclass(frozen=True)
class UnsignedTx:
    """A built-but-unsigned transaction the policy-gated wallet signs. ``instruction_name`` +
    ``program_id`` are exactly what a wallet allow-list checks before it will sign; the
    ``instructions`` tuple may carry a non-spending ComputeBudget prelude."""

    instructions: tuple[Instruction, ...]
    program_id: Pubkey
    instruction_name: str


@dataclass(frozen=True)
class MarketAccount:
    """Decoded ``Market`` state — enough to route ``claim`` to the winning side and print the
    settled outcome. The program never lets us store payloads; this is public on-chain state."""

    fixture_id: int
    stat_key: int
    predicate: TraderPredicate
    vault: str
    stake_yes: int
    stake_no: int
    state: str  # "Open" | "Settled"
    winner: Side  # meaningful only once state == "Settled"


# ── minimal Borsh writer (little-endian, matches Anchor/Borsh exactly) ───────────
class _Borsh:
    def __init__(self) -> None:
        self.buf = bytearray()

    def u8(self, v: int) -> "_Borsh":
        self.buf.append(v & 0xFF)
        return self

    def u32(self, v: int) -> "_Borsh":
        self.buf += int(v).to_bytes(4, "little", signed=False)
        return self

    def i32(self, v: int) -> "_Borsh":
        self.buf += int(v).to_bytes(4, "little", signed=True)
        return self

    def u64(self, v: int) -> "_Borsh":
        self.buf += int(v).to_bytes(8, "little", signed=False)
        return self

    def i64(self, v: int) -> "_Borsh":
        self.buf += int(v).to_bytes(8, "little", signed=True)
        return self

    def arr32(self, v: list[int]) -> "_Borsh":
        if len(v) != 32:
            raise ForgeError(f"expected 32 bytes, got {len(v)}")
        self.buf += bytes(v)
        return self

    def bytes(self) -> bytes:
        return bytes(self.buf)


def _proof_vec(b: _Borsh, nodes: list[dict[str, Any]]) -> None:
    b.u32(len(nodes))
    for n in nodes:
        b.arr32(list(n["hash"])).u8(1 if n["isRightSibling"] else 0)


# ── PDA derivations ──────────────────────────────────────────────────────────────
def market_pda(fixture_id: int, stat_key: int) -> tuple[Pubkey, int]:
    return Pubkey.find_program_address(
        [
            MARKET_SEED,
            fixture_id.to_bytes(8, "little", signed=True),
            stat_key.to_bytes(4, "little"),
        ],
        FORGE_PROGRAM_ID,
    )


def vault_pda(market: Pubkey) -> tuple[Pubkey, int]:
    return Pubkey.find_program_address([VAULT_SEED, bytes(market)], FORGE_PROGRAM_ID)


def position_pda(market: Pubkey, staker: Pubkey) -> tuple[Pubkey, int]:
    return Pubkey.find_program_address(
        [POSITION_SEED, bytes(market), bytes(staker)], FORGE_PROGRAM_ID
    )


def daily_scores_roots_pda(min_timestamp_ms: int) -> tuple[Pubkey, int, int]:
    """The txoracle-owned root account settle passes through. Seed = the epoch-day of the
    snapshot's ``minTimestamp`` (NOT the proof's top-level ``ts``)."""
    epoch_day = min_timestamp_ms // _EPOCH_DAY_MS
    if not 0 <= epoch_day <= 0xFFFF:
        raise ForgeError("epoch day outside u16 range")
    pda, bump = Pubkey.find_program_address(
        [ROOTS_SEED, epoch_day.to_bytes(2, "little")], TXORACLE_PROGRAM_ID
    )
    return pda, bump, epoch_day


def compute_unit_limit_ix(units: int = SETTLE_COMPUTE_UNITS) -> Instruction:
    """ComputeBudget ``SetComputeUnitLimit`` (variant 2) — u32 LE units, no accounts."""
    return Instruction(COMPUTE_BUDGET_ID, bytes([2]) + int(units).to_bytes(4, "little"), [])


# ── instruction builders (account order MUST match each #[derive(Accounts)]) ─────
def build_create_market_ix(
    fixture_id: int, stat_key: int, predicate: TraderPredicate, authority: Pubkey
) -> Instruction:
    market, _ = market_pda(fixture_id, stat_key)
    vault, _ = vault_pda(market)
    data = (
        DISCRIMINATORS["create_market"]
        + _Borsh()
        .i64(fixture_id)
        .u32(stat_key)
        .i32(predicate.threshold)
        .u8(int(predicate.comparison))
        .bytes()
    )
    metas = [
        AccountMeta(market, is_signer=False, is_writable=True),
        AccountMeta(vault, is_signer=False, is_writable=False),
        AccountMeta(authority, is_signer=True, is_writable=True),
        AccountMeta(SYSTEM_PROGRAM_ID, is_signer=False, is_writable=False),
    ]
    return Instruction(FORGE_PROGRAM_ID, data, metas)


def build_stake_ix(
    fixture_id: int, stat_key: int, staker: Pubkey, side: Side, amount_lamports: int
) -> Instruction:
    if amount_lamports <= 0:
        raise ForgeError("stake amount must be positive")
    market, _ = market_pda(fixture_id, stat_key)
    vault, _ = vault_pda(market)
    position, _ = position_pda(market, staker)
    data = DISCRIMINATORS["stake"] + _Borsh().u8(side_index(side)).u64(amount_lamports).bytes()
    metas = [
        AccountMeta(market, is_signer=False, is_writable=True),
        AccountMeta(position, is_signer=False, is_writable=True),
        AccountMeta(vault, is_signer=False, is_writable=True),
        AccountMeta(staker, is_signer=True, is_writable=True),
        AccountMeta(SYSTEM_PROGRAM_ID, is_signer=False, is_writable=False),
    ]
    return Instruction(FORGE_PROGRAM_ID, data, metas)


def encode_settle_args(proof: dict[str, Any]) -> bytes:
    """Borsh-encode the ``settle`` args from a recorded txoracle proof. Mirrors the Rust
    ``data_settle`` order — and OMITS the predicate (the program injects the market's own).

    ``ts`` <- ``summary.updateStats.minTimestamp``. stat_b / op are ``None``."""
    summary = proof["summary"]
    us = summary["updateStats"]
    b = _Borsh()
    b.buf += DISCRIMINATORS["settle"]
    b.i64(us["minTimestamp"])  # ts — the FIX (not proof["ts"])
    # fixture_summary: ScoresBatchSummary
    b.i64(summary["fixtureId"])
    b.i32(us["updateCount"]).i64(us["minTimestamp"]).i64(us["maxTimestamp"])
    b.arr32(list(summary["eventStatsSubTreeRoot"]))
    # fixture_proof <- subTreeProof ; main_tree_proof <- mainTreeProof
    _proof_vec(b, proof["subTreeProof"])
    _proof_vec(b, proof["mainTreeProof"])
    # stat_a: StatTerm = statToProve + eventStatRoot + statProof
    stat = proof["statToProve"]
    b.u32(stat["key"]).i32(stat["value"]).i32(stat["period"])
    b.arr32(list(proof["eventStatRoot"]))
    _proof_vec(b, proof["statProof"])
    # stat_b: Option<StatTerm> = None ; op: Option<BinaryExpression> = None
    b.u8(0).u8(0)
    return b.bytes()


def build_settle_ixs(fixture_id: int, stat_key: int, proof: dict[str, Any]) -> list[Instruction]:
    """[ComputeBudget prelude, settle]. The settle account order mirrors ``Settle``:
    market (w), daily_scores_merkle_roots (r), txoracle_program (r)."""
    market, _ = market_pda(fixture_id, stat_key)
    roots, _, _ = daily_scores_roots_pda(proof["summary"]["updateStats"]["minTimestamp"])
    settle = Instruction(
        FORGE_PROGRAM_ID,
        encode_settle_args(proof),
        [
            AccountMeta(market, is_signer=False, is_writable=True),
            AccountMeta(roots, is_signer=False, is_writable=False),
            AccountMeta(TXORACLE_PROGRAM_ID, is_signer=False, is_writable=False),
        ],
    )
    return [compute_unit_limit_ix(), settle]


def build_claim_ix(fixture_id: int, stat_key: int, staker: Pubkey) -> Instruction:
    market, _ = market_pda(fixture_id, stat_key)
    vault, _ = vault_pda(market)
    position, _ = position_pda(market, staker)
    metas = [
        AccountMeta(market, is_signer=False, is_writable=False),
        AccountMeta(position, is_signer=False, is_writable=True),
        AccountMeta(vault, is_signer=False, is_writable=True),
        AccountMeta(staker, is_signer=True, is_writable=True),
        AccountMeta(SYSTEM_PROGRAM_ID, is_signer=False, is_writable=False),
    ]
    return Instruction(FORGE_PROGRAM_ID, DISCRIMINATORS["claim"], metas)


# ── UnsignedTx wrappers (what the wallet signs; carry the allow-list identity) ────
def create_market_tx(
    fixture_id: int, stat_key: int, predicate: TraderPredicate, authority: Pubkey
) -> UnsignedTx:
    ix = build_create_market_ix(fixture_id, stat_key, predicate, authority)
    return UnsignedTx((ix,), FORGE_PROGRAM_ID, "create_market")


def stake_tx(
    fixture_id: int, stat_key: int, staker: Pubkey, side: Side, amount_lamports: int
) -> UnsignedTx:
    ix = build_stake_ix(fixture_id, stat_key, staker, side, amount_lamports)
    return UnsignedTx((ix,), FORGE_PROGRAM_ID, "stake")


def settle_tx(fixture_id: int, stat_key: int, proof: dict[str, Any]) -> UnsignedTx:
    ixs = build_settle_ixs(fixture_id, stat_key, proof)
    return UnsignedTx(tuple(ixs), FORGE_PROGRAM_ID, "settle")


def claim_tx(fixture_id: int, stat_key: int, staker: Pubkey) -> UnsignedTx:
    ix = build_claim_ix(fixture_id, stat_key, staker)
    return UnsignedTx((ix,), FORGE_PROGRAM_ID, "claim")


# ── Market decode (public on-chain state → routing + display) ─────────────────────
def decode_market(data: bytes) -> MarketAccount:
    """Decode a ``Market`` account (8-byte Anchor disc + Borsh body). Field order mirrors
    ``state.rs`` exactly."""
    if len(data) < 8 + 75:
        raise ForgeError("market account too small to decode")
    o = 8  # skip the Anchor discriminator
    fixture_id = int.from_bytes(data[o : o + 8], "little", signed=True)
    stat_key = int.from_bytes(data[o + 8 : o + 12], "little")
    threshold = int.from_bytes(data[o + 12 : o + 16], "little", signed=True)
    comparison = Comparison(data[o + 16])
    vault = str(Pubkey.from_bytes(data[o + 17 : o + 49]))
    stake_yes = int.from_bytes(data[o + 49 : o + 57], "little")
    stake_no = int.from_bytes(data[o + 57 : o + 65], "little")
    state = "Open" if data[o + 65] == 0 else "Settled"
    winner: Side = "Yes" if data[o + 66] == 0 else "No"
    return MarketAccount(
        fixture_id=fixture_id,
        stat_key=stat_key,
        predicate=TraderPredicate(threshold, comparison),
        vault=vault,
        stake_yes=stake_yes,
        stake_no=stake_no,
        state=state,
        winner=winner,
    )


# ── recorded proof helpers ────────────────────────────────────────────────────────
def load_recorded_proof(path: str | Path) -> dict[str, Any]:
    """Load the recorded settled-fixture proof (the free World-Cup tier). Returns the inner
    ``proof`` object."""
    doc = json.loads(Path(path).read_text())
    proof: dict[str, Any] = doc.get("proof", doc)
    return proof


def winning_predicate(proof: dict[str, Any]) -> TraderPredicate:
    """A predicate that HOLDS for this proof: ``stat.value > (value - 1)`` — always true —
    so a valid settle records ``winner = Yes`` (the demo's YES-wins path)."""
    value = int(proof["statToProve"]["value"])
    return TraderPredicate(threshold=value - 1, comparison=Comparison.GREATER_THAN)


# Human names for the program's custom errors (code = 6000 + variant index), so a live
# failure reports the exact fail-closed reason instead of a bare number.
SETTLEMENT_ERRORS: dict[int, str] = {
    6000: "MarketNotOpen",
    6001: "MarketNotSettled",
    6002: "ZeroStake",
    6003: "NotWinningSide",
    6004: "AlreadyClaimed",
    6005: "NoWinningStake",
    6006: "WrongOracleProgram",
    6007: "Overflow",
    6008: "OracleNoReturnData",
    6009: "OracleReturnWrongProgram",
    6010: "OracleBadReturnData",
}

File: frontend/lib/solana/forge-client.ts
import {
  PublicKey,
  SystemProgram,
  TransactionInstruction,
} from "@solana/web3.js";
import { Buffer } from "buffer";

import { ByteReader, ByteWriter } from "./borsh";
import { FORGE_PROGRAM_ID } from "./config";

/**
 * Client mirror of `backend/gorilla/forge_client.py`. Turns an intent into
 * the EXACT `forge_markets` instruction the deployed devnet program expects —
 * right 8-byte Anchor discriminator, right account metas, right Borsh args, right
 * PDA derivations — and decodes the public on-chain `Market`/`Position` state.
 * The program is frozen; this file mirrors its wire format as DATA.
 *
 * The read + `stake` + `create_market` + `claim` paths are ported here (what the
 * web surface needs — browser stakers place bets AND collect payouts); `settle`
 * stays server-side in the backend (it carries the oracle proof).
 */

export type Side = "Yes" | "No";
export type MarketStateName = "Open" | "Settled";

export enum Comparison {
  GreaterThan = 0,
  LessThan = 1,
  EqualTo = 2,
}

export const COMPARISON_SYMBOL: Record<Comparison, string> = {
  [Comparison.GreaterThan]: ">",
  [Comparison.LessThan]: "<",
  [Comparison.EqualTo]: "=",
};

// Borsh variant index for the program's `Side` enum (Yes = 0, No = 1).
const SIDE_INDEX: Record<Side, number> = { Yes: 0, No: 1 };

// Anchor discriminators = sha256("global:<ix>")[..8] — verified against the IDL.
export const DISCRIMINATORS: Record<string, number[]> = {
  create_market: [103, 226, 97, 235, 200, 188, 251, 254],
  stake: [206, 176, 202, 18, 200, 209, 179, 108],
  settle: [175, 42, 185, 87, 144, 131, 102, 212],
  claim: [62, 198, 214, 193, 213, 159, 108, 210],
};

// Program custom errors (code = 6000 + variant index) — so a simulated stake that
// fails closed reports the exact reason, not a bare number.
export const SETTLEMENT_ERRORS: Record<number, string> = {
  6000: "MarketNotOpen",
  6001: "MarketNotSettled",
  6002: "ZeroStake",
  6003: "NotWinningSide",
  6004: "AlreadyClaimed",
  6005: "NoWinningStake",
  6006: "WrongOracleProgram",
  6007: "Overflow",
  6008: "OracleNoReturnData",
  6009: "OracleReturnWrongProgram",
  6010: "OracleBadReturnData",
  6011: "FixtureMismatch",
  6012: "StatMismatch",
  6013: "MultiStatNotAllowed",
  6014: "PeriodMismatch",
  6015: "WrongEngineProgram",
  6016: "MarketLocked",
  6017: "ReclaimTooEarly",
  6018: "ReclaimUnavailable",
};

export function toLamports(sol: number): bigint {
  return BigInt(Math.round(sol * 1e9));
}

export function lamportsToSol(lamports: bigint | number): number {
  return Number(lamports) / 1e9;
}

// ── PDA seeds (mirror forge-markets/src/interface.rs) ──────────────────────────
const enc = new TextEncoder();
const MARKET_SEED = enc.encode("market");
const VAULT_SEED = enc.encode("vault");
const POSITION_SEED = enc.encode("position");

export function marketPda(fixtureId: bigint, statKey: number): [PublicKey, number] {
  const seed = new ByteWriter().i64(fixtureId).build();
  const key = new ByteWriter().u32(statKey).build();
  return PublicKey.findProgramAddressSync(
    [MARKET_SEED, seed, key],
    FORGE_PROGRAM_ID,
  );
}

export function vaultPda(market: PublicKey): [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [VAULT_SEED, market.toBytes()],
    FORGE_PROGRAM_ID,
  );
}

export function positionPda(
  market: PublicKey,
  staker: PublicKey,
): [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [POSITION_SEED, market.toBytes(), staker.toBytes()],
    FORGE_PROGRAM_ID,
  );
}

// ── Decoded account shapes (public on-chain state only) ────────────────────────
export interface TraderPredicate {
  threshold: number;
  comparison: Comparison;
}

export interface MarketAccount {
  address: string;
  fixtureId: bigint;
  statKey: number;
  predicate: TraderPredicate;
  vault: string;
  stakeYes: bigint;
  stakeNo: bigint;
  state: MarketStateName;
  winner: Side; // meaningful only once state === "Settled"
  authority: string;
  potLamports: bigint;
}

export interface PositionAccount {
  address: string;
  market: string;
  owner: string;
  side: Side;
  amount: bigint;
  claimed: boolean;
}

// Byte size of each account (8-byte disc + Borsh body) — used as a getProgramAccounts
// dataSize filter, so we never need the account discriminator.
export const MARKET_ACCOUNT_SIZE = 142;
export const POSITION_ACCOUNT_SIZE = 99;

/** Decode a `Market` account. Field order mirrors `state.rs` exactly. */
export function decodeMarket(address: string, data: Uint8Array): MarketAccount {
  if (data.length < MARKET_ACCOUNT_SIZE) {
    throw new Error("market account too small to decode");
  }
  const r = new ByteReader(data, 8); // skip the Anchor discriminator
  const fixtureId = r.i64();
  const statKey = r.u32();
  const threshold = r.i32();
  const comparison = r.u8() as Comparison;
  const vault = new PublicKey(r.fixed(32)).toBase58();
  const stakeYes = r.u64();
  const stakeNo = r.u64();
  const state: MarketStateName = r.u8() === 0 ? "Open" : "Settled";
  const winner: Side = r.u8() === 0 ? "Yes" : "No";
  const authority = new PublicKey(r.fixed(32)).toBase58();
  return {
    address,
    fixtureId,
    statKey,
    predicate: { threshold, comparison },
    vault,
    stakeYes,
    stakeNo,
    state,
    winner,
    authority,
    potLamports: stakeYes + stakeNo,
  };
}

/** Decode a `Position` account. Field order mirrors `state.rs` exactly. */
export function decodePosition(address: string, data: Uint8Array): PositionAccount {
  if (data.length < POSITION_ACCOUNT_SIZE) {
    throw new Error("position account too small to decode");
  }
  const r = new ByteReader(data, 8);
  const market = new PublicKey(r.fixed(32)).toBase58();
  const owner = new PublicKey(r.fixed(32)).toBase58();
  const side: Side = r.u8() === 0 ? "Yes" : "No";
  const amount = r.u64();
  const claimed = r.bool();
  return { address, market, owner, side, amount, claimed };
}

/**
 * Build the `stake` instruction — the interactive place-a-bet action.
 * Account order MUST match `#[derive(Accounts)] Stake`:
 *   market (w), position (w), vault (w), staker (signer, w), system (r).
 * data = disc("stake") + u8(side) + u64(amount).
 */
export function buildStakeIx(params: {
  fixtureId: bigint;
  statKey: number;
  staker: PublicKey;
  side: Side;
  amountLamports: bigint;
}): { instruction: TransactionInstruction; market: PublicKey; position: PublicKey; vault: PublicKey } {
  const { fixtureId, statKey, staker, side, amountLamports } = params;
  if (amountLamports <= 0n) throw new Error("stake amount must be positive");
  const [market] = marketPda(fixtureId, statKey);
  const [vault] = vaultPda(market);
  const [position] = positionPda(market, staker);

  const data = new ByteWriter()
    .bytes(DISCRIMINATORS.stake)
    .u8(SIDE_INDEX[side])
    .u64(amountLamports)
    .build();

  const instruction = new TransactionInstruction({
    programId: FORGE_PROGRAM_ID,
    keys: [
      { pubkey: market, isSigner: false, isWritable: true },
      { pubkey: position, isSigner: false, isWritable: true },
      { pubkey: vault, isSigner: false, isWritable: true },
      { pubkey: staker, isSigner: true, isWritable: true },
      { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
    ],
    data: Buffer.from(data),
  });

  return { instruction, market, position, vault };
}

/**
 * Build the `create_market` instruction — open a two-sided market over
 * `(fixture_id, stat_key, period)`. Account order MUST match
 * `#[derive(Accounts)] CreateMarket` (see program/src/instructions/create_market.rs):
 *   market (w), vault (r), authority (signer, w), system (r).
 * The vault is NOT created here (it's a system-owned PDA that first receives
 * lamports on stake), so it is read-only — Anchor only validates the derivation.
 * data = disc("create_market") + i64(fixture) + u32(stat) + i32(threshold)
 *      + u8(comparison) + i32(period) + i64(lockTs), mirroring backend
 *      `build_create_market_ix`. `lockTs` is the betting cutoff (Unix seconds);
 *      pass 0n for no cutoff — see program stake.rs / create_market.rs (#36).
 */
export function buildCreateMarketIx(params: {
  fixtureId: bigint;
  statKey: number;
  threshold: number;
  comparison: Comparison;
  period: number;
  lockTs: bigint;
  authority: PublicKey;
}): { instruction: TransactionInstruction; market: PublicKey; vault: PublicKey } {
  const { fixtureId, statKey, threshold, comparison, period, lockTs, authority } = params;
  const [market] = marketPda(fixtureId, statKey);
  const [vault] = vaultPda(market);

  const data = new ByteWriter()
    .bytes(DISCRIMINATORS.create_market)
    .i64(fixtureId)
    .u32(statKey)
    .i32(threshold)
    .u8(comparison)
    .i32(period)
    .i64(lockTs)
    .build();

  const instruction = new TransactionInstruction({
    programId: FORGE_PROGRAM_ID,
    keys: [
      { pubkey: market, isSigner: false, isWritable: true },
      { pubkey: vault, isSigner: false, isWritable: false },
      { pubkey: authority, isSigner: true, isWritable: true },
      { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
    ],
    data: Buffer.from(data),
  });

  return { instruction, market, vault };
}

/**
 * Build the `claim` instruction — a winning staker withdraws their pro-rata share.
 * Layout cross-checked against BOTH the backend (`build_claim_ix`) and the Rust
 * context (`instructions/claim.rs::Claim`), which agree. Account order MUST match
 * `#[derive(Accounts)] Claim`:
 *   market (r), position (w), vault (w), staker (signer, w), system (r).
 * The market is READ-ONLY here (unlike stake): claim only reads the settled state;
 * lamports leave the vault, and the position's `claimed` flag flips.
 * data = disc("claim") — the instruction takes no args.
 */
export function buildClaimIx(params: {
  fixtureId: bigint;
  statKey: number;
  staker: PublicKey;
}): { instruction: TransactionInstruction; market: PublicKey; position: PublicKey; vault: PublicKey } {
  const { fixtureId, statKey, staker } = params;
  const [market] = marketPda(fixtureId, statKey);
  const [vault] = vaultPda(market);
  const [position] = positionPda(market, staker);

  const instruction = new TransactionInstruction({
    programId: FORGE_PROGRAM_ID,
    keys: [
      { pubkey: market, isSigner: false, isWritable: false },
      { pubkey: position, isSigner: false, isWritable: true },
      { pubkey: vault, isSigner: false, isWritable: true },
      { pubkey: staker, isSigner: true, isWritable: true },
      { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
    ],
    data: Buffer.from(DISCRIMINATORS.claim),
  });

  return { instruction, market, position, vault };
}

/** Map an Anchor custom-error code from a simulation into its program error name. */
export function settlementErrorName(code: number): string | null {
  return SETTLEMENT_ERRORS[code] ?? null;
}

/** Extract an Anchor custom-error code from a simulation/transaction err object,
 * e.g. `{ InstructionError: [0, { Custom: 6000 }] }` → `6000`. */
export function customErrorCode(err: unknown): number | null {
  const instructionError = (err as { InstructionError?: [number, unknown] })
    ?.InstructionError;
  if (
    Array.isArray(instructionError) &&
    instructionError[1] &&
    typeof instructionError[1] === "object" &&
    "Custom" in instructionError[1]
  ) {
    return (instructionError[1] as { Custom: number }).Custom;
  }
  return null;
}
